env.bzl

```diff
@@ -13,28 +13,53 @@ py_runtime(
 )
 """
 
+def _label_from_condabin(rctx, exe_name):
+    return Label("@{}//:{}/condabin/{}{}".format(
+        rctx.attr.conda_repo,
+        rctx.attr.conda_dir,
+        exe_name,
+        CONDA_EXT_MAP[get_os(rctx)],
+    ))
+
+def _user_chosen_executable(rctx):
+    if rctx.attr.use_mamba:
+        return _label_from_condabin(rctx, "mamba")
+    return _label_from_condabin(rctx, "conda")
+
 # clean conda caches and unused packages
 def _clean(rctx, executable):
     rctx.report_progress("Cleaning up")
 
-    args = [rctx.path(executable), "clean", "-a", "-y"]
+    args = [executable, "clean", "-a", "-y"]
 
     result = rctx.execute(args, quiet = rctx.attr.quiet, timeout = rctx.attr.timeout)
     if result.return_code:
         fail("Failure cleaning up.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
 
+def _create_empty_environment(rctx, executable, env_name):
+    rctx.report_progress("Creating empty conda environment, to be populated afterwards")
+    args = [executable, "create", "-y", "-p", "./{}".format(env_name)]
+    result = rctx.execute(args, quiet = rctx.attr.quiet, timeout = rctx.attr.timeout)
+    if result.return_code:
+        fail("Failure creating empty environment.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
+
+def _update_environment(rctx, executable, env_name, env_file):
+    rctx.report_progress("Updating empty conda environment to populate it")
+    args = [executable, "env", "update", "-f", env_file, "-p", "./{}".format(env_name)]
+    result = rctx.execute(args, quiet = rctx.attr.quiet, timeout = rctx.attr.timeout)
+    if result.return_code:
+        fail("Failure updating environment.\nstdout: {}\nstderr: {}".format(result.stdout, result.stderr))
+
 # create new local conda environment from file
 def _create_environment(rctx, executable, env_name):
     rctx.report_progress("Creating conda environment")
 
+    executable = rctx.path(_user_chosen_executable(rctx))
     # path to env file as string
     env_file = str(rctx.path(rctx.attr.environment))
```

Comment:
Maybe casing to string is not necessary also here? Could you try and test and remove if it works?

Reply:
And additionally I just spotted that `executable` and `env_file` are already passed to this function